extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java

```diff
@@ -0,0 +1,192 @@
+package io.quarkus.narayana.jta.runtime.internal.tsr;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import jakarta.transaction.Status;
+import jakarta.transaction.Synchronization;
+import jakarta.transaction.SystemException;
+import jakarta.transaction.TransactionSynchronizationRegistry;
+
+import org.jboss.logging.Logger;
+
+public class AgroalOrderedLastSynchronizationList implements Synchronization {
+    private static final Logger LOGGER = Logger.getLogger(AgroalOrderedLastSynchronizationList.class);
+    private static final String ADD_SYNC_ERROR = "Syncs are not allowed because the group of synchronizations to which this sync belongs has already ran";
+    private static final String REGISTER_SYNC_ERROR = "Syncs are not allowed to be registered when the transaction is in state ";
+
+    // order the groups of synchronization as follows (Agroal is last since it needs to validate that
+    // connection wrappers are closed at the right time):
+    private static final String[] PKG_PREFIXES = { "", "org.hibernate", "io.agroal.narayana" };
+
+    /*
+     * Keep track of whether a synchronization group has been processed.
+     * If a group of synchs has already been processed then do not allow further synchs to be registered in that group.
+     * If a group of synchs is currently being processed then allow it to be registered.
+     * But note that no synchronizations can be registered after the transaction has finished preparing.
+     */
+    private enum ExecutionStatus {
+        PENDING, // the synchronization has not been called yet
+        RUNNING, // the synchronization is execution
+        FINISHED // the synchronization has executed
+    }
+
+    /*
+     * Synchronizations are grouped by package prefix and these groups are ordered such that the
+     * synchronizations in the first group execute first, then the second group is processed, etc.
+     * In particular, the Agroal synchronization group runs last.
+     *
+     * The beforeCompletion methods within a group are called in the order they were added,
+     * and the afterCompletion methods are ran in the reverse order
+     */
+    private class SynchronizationGroup implements Synchronization {
+        String packagePrefix; // Synchronizations with this package prefix belong to this group
+        List<Synchronization> synchs; // the Synchronizations in the group
+        ExecutionStatus status; // track the status to decide when it's too late to allow more registrations
```

Comment:
Should `status` be volatile so it doesn't get placed in a CPU register (or whatever thread memory is considered to be) or read from a CPU register?

Reply:
Agreed